# Computational savings analysis (ROMS/MARBL)

Raw timing data extracted from `MPI_run_time` / `ntimes` in the `output.log` files under `roms_marbl_dic/exp*` (DOR-forcing baseline), `roms_marbl_alk/exp*` (OAE-forcing baseline), and `roms/exp*` (DOR- and OAE-forced runs with 2x12 embedded mCDR response tracers, `ANTITRACER` config).

Only runs with a completed `output.log` (i.e. a recorded `MPI_run_time`) are listed. Directories with no output at all (run never completed, or logs since removed) are omitted entirely rather than tracked as rows.

All runs used 7 nodes x 128 cores = 896 cores on Perlmutter, dt = 900 s.

`simulated_days = ntimes * dt_sec / 86400` is computed exactly per run (it varies a little in principle, e.g. a run that stops early), and

`core_hours_per_sim_year = cores * mpi_run_time_sec * (365 / simulated_days) / 3600`

rather than assuming every baseline run is exactly 2 months and every tracer run exactly 24 months.

In [1]:
import pandas as pd

CORES = 7 * 128
DAYS_PER_YEAR = 365.0

raw = [
    # group, site, log_path, machine, ntimes, dt_sec, mpi_run_time_sec, included, exclusion_reason
    ("baseline_dor", "BC10", "roms_marbl_dic/expBC10/OUTPUT/output.log", "perlmutter", 5856, 900, 18907.710334290001, True, None),
    ("baseline_dor", "BC7",  "roms_marbl_dic/expBC7/OUTPUT/output.log",  "perlmutter", 5856, 900, 18407.189070027001, True, None),
    ("baseline_dor", "BC8",  "roms_marbl_dic/expBC8/OUTPUT/output.log",  "perlmutter", 5856, 900, 18495.149602092002, True, None),
    ("baseline_dor", "EC10", "roms_marbl_dic/expEC10/OUTPUT/output.log", "perlmutter", 5856, 900, 18945.762408160001, True, None),
    ("baseline_dor", "EC7",  "roms_marbl_dic/expEC7/OUTPUT/output.log",  "perlmutter", 5856, 900, 18652.749636027002, True, None),
    ("baseline_dor", "EC8",  "roms_marbl_dic/expEC8/OUTPUT/output.log",  "perlmutter", 5856, 900, 18663.107204344000, True, None),
    ("baseline_dor", "JP10", "roms_marbl_dic/expJP10/OUTPUT/output.log", "perlmutter", 5856, 900, 18443.350212056001, True, None),
    ("baseline_dor", "JP7",  "roms_marbl_dic/expJP7/OUTPUT/output.log",  "perlmutter", 5856, 900, 18495.567425382000, True, None),
    ("baseline_dor", "JP8",  "roms_marbl_dic/expJP8/OUTPUT/output.log",  "perlmutter", 5856, 900, 18424.221101873998, True, None),
    ("baseline_dor", "VI10", "roms_marbl_dic/expVI10/OUTPUT/output.log", "perlmutter", 5856, 900, 18464.383533058000, True, None),
    ("baseline_dor", "VI7",  "roms_marbl_dic/expVI7/OUTPUT/output.log",  "perlmutter", 5856, 900, 18490.079755885003, True, None),
    ("baseline_dor", "VI8",  "roms_marbl_dic/expVI8/OUTPUT/output.log",  "perlmutter", 5856, 900, 18473.145958665002, True, None),

    ("baseline_oae", "BC8", "roms_marbl_alk/expBC8_perlmutter/OUTPUT/output.log", "perlmutter", 5856, 900, 18433.289086269000, True, None),
    ("baseline_oae", "EC8", "roms_marbl_alk/expEC8_perlmutter/OUTPUT/output.log", "perlmutter", 5856, 900, 18529.492123329997, True, None),
    ("baseline_oae", "JP8", "roms_marbl_alk/expJP8_perlmutter/OUTPUT/output.log", "perlmutter", 5856, 900, 19099.852160424998, True, None),
    ("baseline_oae", "VI8", "roms_marbl_alk/expVI8_perlmutter/OUTPUT/output.log", "perlmutter", 5856, 900, 18503.861234453001, True, None),

    ("cdr_tracer_dor", "exp0", "roms/exp0/OUTPUT_dor/output.log", "perlmutter", 70368, 900, 131931.84242529800, True, None),
    ("cdr_tracer_dor", "exp1", "roms/exp1/OUTPUT_dor/output.log", "perlmutter", 70368, 900, 131882.78473718200, True, None),
    ("cdr_tracer_dor", "exp2", "roms/exp2/OUTPUT_dor/output.log", "perlmutter", 70368, 900, 131674.65698570499, True, None),
    ("cdr_tracer_dor", "exp6", "roms/exp6/OUTPUT_dor/output.log", "perlmutter", 70368, 900, 131651.77466204800, True, None),

    ("cdr_tracer_oae", "exp2", "roms/exp2/OUTPUT_oae/output.log", "perlmutter", 70368, 900, 150644.64924403001, True, None),
    ("cdr_tracer_oae", "exp3", "roms/exp3/OUTPUT_oae/output.log", "perlmutter", 70368, 900, 151116.97527306300, True, None),
]

df = pd.DataFrame(raw, columns=[
    "group", "site", "log_path", "machine", "ntimes", "dt_sec",
    "mpi_run_time_sec", "included", "exclusion_reason",
])
df["simulated_days"] = df["ntimes"] * df["dt_sec"] / 86400
df

,group,site,log_path,machine,ntimes,dt_sec,mpi_run_time_sec,included,exclusion_reason,simulated_days
0,baseline_dor,BC10,roms_marbl_dic/expBC10/OUTPUT/output.log,perlmutter,5856,900,18907.710334,True,None,61.0
1,baseline_dor,BC7,roms_marbl_dic/expBC7/OUTPUT/output.log,perlmutter,5856,900,18407.189070,True,None,61.0
2,baseline_dor,BC8,roms_marbl_dic/expBC8/OUTPUT/output.log,perlmutter,5856,900,18495.149602,True,None,61.0
3,baseline_dor,EC10,roms_marbl_dic/expEC10/OUTPUT/output.log,perlmutter,5856,900,18945.762408,True,None,61.0
4,baseline_dor,EC7,roms_marbl_dic/expEC7/OUTPUT/output.log,perlmutter,5856,900,18652.749636,True,None,61.0
5,baseline_dor,EC8,roms_marbl_dic/expEC8/OUTPUT/output.log,perlmutter,5856,900,18663.107204,True,None,61.0
6,baseline_dor,JP10,roms_marbl_dic/expJP10/OUTPUT/output.log,perlmutter,5856,900,18443.350212,True,None,61.0
7,baseline_dor,JP7,roms_marbl_dic/expJP7/OUTPUT/output.log,perlmutter,5856,900,18495.567425,True,None,61.0
8,baseline_dor,JP8,roms_marbl_dic/expJP8/OUTPUT/output.log,perlmutter,5856,900,18424.221102,True,None,61.0
9,baseline_dor,VI10,roms_marbl_dic/expVI10/OUTPUT/output.log,perlmutter,5856,900,18464.383533,True,None,61.0


## Per-run core-hours/sim-year

Uses each run's own `simulated_days` rather than an assumed 2-month / 24-month duration, so it's robust to runs of slightly different length.

In [2]:
df["core_hours_per_sim_year"] = (
    CORES * df["mpi_run_time_sec"] * (DAYS_PER_YEAR / df["simulated_days"]) / 3600
)
df

,group,site,log_path,machine,ntimes,dt_sec,mpi_run_time_sec,included,exclusion_reason,simulated_days,core_hours_per_sim_year
0,baseline_dor,BC10,roms_marbl_dic/expBC10/OUTPUT/output.log,perlmutter,5856,900,18907.710334,True,None,61.0,28158.367886
1,baseline_dor,BC7,roms_marbl_dic/expBC7/OUTPUT/output.log,perlmutter,5856,900,18407.189070,True,None,61.0,27412.964998
2,baseline_dor,BC8,roms_marbl_dic/expBC8/OUTPUT/output.log,perlmutter,5856,900,18495.149602,True,None,61.0,27543.960500
3,baseline_dor,EC10,roms_marbl_dic/expEC10/OUTPUT/output.log,perlmutter,5856,900,18945.762408,True,None,61.0,28215.037058
4,baseline_dor,EC7,roms_marbl_dic/expEC7/OUTPUT/output.log,perlmutter,5856,900,18652.749636,True,None,61.0,27778.666853
5,baseline_dor,EC8,roms_marbl_dic/expEC8/OUTPUT/output.log,perlmutter,5856,900,18663.107204,True,None,61.0,27794.091895
6,baseline_dor,JP10,roms_marbl_dic/expJP10/OUTPUT/output.log,perlmutter,5856,900,18443.350212,True,None,61.0,27466.818094
7,baseline_dor,JP7,roms_marbl_dic/expJP7/OUTPUT/output.log,perlmutter,5856,900,18495.567425,True,None,61.0,27544.582745
8,baseline_dor,JP8,roms_marbl_dic/expJP8/OUTPUT/output.log,perlmutter,5856,900,18424.221102,True,None,61.0,27438.330005
9,baseline_dor,VI10,roms_marbl_dic/expVI10/OUTPUT/output.log,perlmutter,5856,900,18464.383533,True,None,61.0,27498.142034


## Group summary and final table

Baseline (DOR + OAE forcing) is pooled into a single average, since forcing type barely affects baseline cost.
DOR- and OAE-forced tracer runs are kept as separate rows, since the ~14% cost gap between them is systematic, not noise.

In [3]:
included = df[df["included"]]

baseline = included[included["group"].isin(["baseline_dor", "baseline_oae"])]
cdr_tracer_dor = included[included["group"] == "cdr_tracer_dor"]
cdr_tracer_oae = included[included["group"] == "cdr_tracer_oae"]

N_BASELINE_EXPERIMENTS = 12  # one per site, matches roms_marbl_dic site count

summary_rows = []

ch_yr = baseline["core_hours_per_sim_year"].mean()
summary_rows.append({
    "configuration": "ROMS/MARBL (baseline)",
    "n_samples": len(baseline),
    "core_hours_per_sim_year": ch_yr,
    "n_experiments": N_BASELINE_EXPERIMENTS,
    "total_core_hours": ch_yr * N_BASELINE_EXPERIMENTS,
    "speedup_vs_baseline": None,
})
baseline_total = summary_rows[0]["total_core_hours"]

for label, sub in [
    ("ROMS DOR + 2x12 mCDR response tracers", cdr_tracer_dor),
    ("ROMS OAE + 2x12 mCDR response tracers", cdr_tracer_oae),
]:
    ch_yr = sub["core_hours_per_sim_year"].mean()
    total = ch_yr * 1
    summary_rows.append({
        "configuration": label,
        "n_samples": len(sub),
        "core_hours_per_sim_year": ch_yr,
        "n_experiments": 1,
        "total_core_hours": total,
        "speedup_vs_baseline": baseline_total / total,
    })

summary = pd.DataFrame(summary_rows).round(1)
summary

,configuration,n_samples,core_hours_per_sim_year,n_experiments,total_core_hours,speedup_vs_baseline
0,ROMS/MARBL (baseline),16,27684.2,12,332210.2,NaN
1,ROMS DOR + 2x12 mCDR response tracers,4,16332.8,1,16332.8,20.3
2,ROMS OAE + 2x12 mCDR response tracers,2,18699.4,1,18699.4,17.8


# Computational savings analysis (CESM/MARBL)

Timing data for CESM comes from the standard `timing/cesm_timing.*` files CIME writes per job submission, under `/global/cfs/projectdirs/m4746/Users/nora/Ocean-CDR-Atlas-v0/cases/`. Unlike the ROMS `output.log`, these already report `Model Cost` directly in **pe-hrs/simulated_year** (pe = MPI task, 896 total, same 7x128-core layout as ROMS), so no manual annualization is needed.

- **Baseline**: two full-BGC samples, both `stop_n=1` (1 simulated year), representing the cost of one naive per-tracer/per-site experiment without the embedded-tracer scheme:
  - `smyle.cdr-atlas-v0.control.001_backup` -- a single continuous 16-year run (5840 days, `stop_n=16`), no CDR forcing.
  - `smyle.cdr-atlas-v0.glb-dor_North_Atlantic_basin_000_1999-01-01_00000.001` -- a real single-site DOR-forced run (`dic_forcing_scale_factor=-1.0`, no `ANTITRACER`/embedded-tracer scheme), i.e. exactly the kind of naive per-site experiment the embedded-tracer approach replaces. (`smyle.cdr-atlas-v0.control.001`, without `_backup`, has no completed timing data yet.)
- **DOR tracer run**: `smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001` -- a multi-year run submitted as 15 separate 1-year job segments (`stop_n=1` each, `continue_run=TRUE` after the first), each with its own timing file.
- **OAE tracer run**: `smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001` -- same structure, 15 yearly segments.

As with ROMS, DOR- and OAE-forced tracer runs are kept as separate rows: OAE is consistently about 2.1x the cost of DOR here (vs. ~1.14x for ROMS), so pooling them would hide a real, systematic difference.

In [4]:
# group, case/segment id, log_path, pe_count, model_cost_pe_hrs_per_sim_year
cesm_raw = [
    ("baseline", "control_backup.47708726.260112-145931",
     "smyle.cdr-atlas-v0.control.001_backup/timing/", 896, 446.29),
    ("baseline", "glb-dor_North_Atlantic_basin_000.40893496.250719-212253",
     "smyle.cdr-atlas-v0.glb-dor_North_Atlantic_basin_000_1999-01-01_00000.001/timing/", 896, 399.22),

    ("cdr_tracer_dor", "52615295.260511-074229", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 11061.81),
    ("cdr_tracer_dor", "52847077.260512-043929", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 11126.96),
    ("cdr_tracer_dor", "52895706.260513-185359", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10905.42),
    ("cdr_tracer_dor", "53024860.260516-065442", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10821.49),
    ("cdr_tracer_dor", "53057391.260517-084256", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 11585.07),
    ("cdr_tracer_dor", "53098962.260520-212010", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10681.59),
    ("cdr_tracer_dor", "53533646.260529-072156", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 11572.97),
    ("cdr_tracer_dor", "53717751.260603-082239", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10613.12),
    ("cdr_tracer_dor", "53907440.260604-022006", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10822.14),
    ("cdr_tracer_dor", "54022125.260605-141321", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10726.33),
    ("cdr_tracer_dor", "54042030.260606-061620", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10657.67),
    ("cdr_tracer_dor", "55421638.260703-075818", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 12288.26),
    ("cdr_tracer_dor", "55481590.260704-154939", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10900.88),
    ("cdr_tracer_dor", "55525269.260705-082608", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10602.52),
    ("cdr_tracer_dor", "55565201.260706-063326", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-dor-daily.001/timing/", 896, 10557.34),

    ("cdr_tracer_oae", "50496031.260324-214811", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22789.32),
    ("cdr_tracer_oae", "50567117.260326-221801", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22875.48),
    ("cdr_tracer_oae", "50645807.260328-102848", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22826.76),
    ("cdr_tracer_oae", "50691751.260329-135008", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22893.10),
    ("cdr_tracer_oae", "50743372.260330-230524", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22939.64),
    ("cdr_tracer_oae", "50817938.260401-093403", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 23180.65),
    ("cdr_tracer_oae", "50933767.260403-082210", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22908.52),
    ("cdr_tracer_oae", "51013817.260405-002834", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22990.62),
    ("cdr_tracer_oae", "51088262.260406-065847", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 23153.31),
    ("cdr_tracer_oae", "51186406.260408-004236", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22870.14),
    ("cdr_tracer_oae", "51382757.260415-154123", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22667.04),
    ("cdr_tracer_oae", "51683356.260417-184545", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22915.19),
    ("cdr_tracer_oae", "51901706.260422-235411", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22806.08),
    ("cdr_tracer_oae", "52033732.260425-080118", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22732.60),
    ("cdr_tracer_oae", "52091157.260426-131554", "smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-oae-daily.001/timing/", 896, 22807.43),
]

cesm_df = pd.DataFrame(cesm_raw, columns=[
    "group", "lid", "case_path", "pe_count", "core_hours_per_sim_year",
])
cesm_df

,group,lid,case_path,pe_count,core_hours_per_sim_year
0,baseline,control_backup.47708726.260112-145931,smyle.cdr-atlas-v0.control.001_backup/timing/,896,446.29
1,baseline,glb-dor_North_Atlantic_basin_000.40893496.2507...,smyle.cdr-atlas-v0.glb-dor_North_Atlantic_basi...,896,399.22
2,cdr_tracer_dor,52615295.260511-074229,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,11061.81
3,cdr_tracer_dor,52847077.260512-043929,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,11126.96
4,cdr_tracer_dor,52895706.260513-185359,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,10905.42
5,cdr_tracer_dor,53024860.260516-065442,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,10821.49
6,cdr_tracer_dor,53057391.260517-084256,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,11585.07
7,cdr_tracer_dor,53098962.260520-212010,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,10681.59
8,cdr_tracer_dor,53533646.260529-072156,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,11572.97
9,cdr_tracer_dor,53717751.260603-082239,smyle.cdr-atlas-v0.glb-antitracer_1999-01_all-...,896,10613.12


## CESM group summary and final table

Baseline is a single measured 16-year run (`control.001_backup`), not your colleague's original 483 estimate. DOR and OAE tracer segments are each averaged over their 15 yearly job chunks and kept as separate rows.

In [5]:
cesm_baseline = cesm_df[cesm_df["group"] == "baseline"]
cesm_tracer_dor = cesm_df[cesm_df["group"] == "cdr_tracer_dor"]
cesm_tracer_oae = cesm_df[cesm_df["group"] == "cdr_tracer_oae"]

N_CESM_BASELINE_EXPERIMENTS = 690  # one per response tracer

cesm_summary_rows = []

ch_yr = cesm_baseline["core_hours_per_sim_year"].mean()
cesm_summary_rows.append({
    "configuration": "CESM/MARBL (baseline)",
    "n_samples": len(cesm_baseline),
    "core_hours_per_sim_year": ch_yr,
    "n_experiments": N_CESM_BASELINE_EXPERIMENTS,
    "total_core_hours": ch_yr * N_CESM_BASELINE_EXPERIMENTS,
    "speedup_vs_baseline": None,
})
cesm_baseline_total = cesm_summary_rows[0]["total_core_hours"]

for label, sub in [
    ("CESM DOR + 1x690 mCDR response tracers", cesm_tracer_dor),
    ("CESM OAE + 1x690 mCDR response tracers", cesm_tracer_oae),
]:
    ch_yr = sub["core_hours_per_sim_year"].mean()
    total = ch_yr * 1
    cesm_summary_rows.append({
        "configuration": label,
        "n_samples": len(sub),
        "core_hours_per_sim_year": ch_yr,
        "n_experiments": 1,
        "total_core_hours": total,
        "speedup_vs_baseline": cesm_baseline_total / total,
    })

cesm_summary = pd.DataFrame(cesm_summary_rows).round(1)
cesm_summary

,configuration,n_samples,core_hours_per_sim_year,n_experiments,total_core_hours,speedup_vs_baseline
0,CESM/MARBL (baseline),2,422.8,690,291701.0,NaN
1,CESM DOR + 1x690 mCDR response tracers,15,10994.9,1,10994.9,26.5
2,CESM OAE + 1x690 mCDR response tracers,15,22890.4,1,22890.4,12.7


## Combined final table

| Configuration | Core-hours per sim year | # Experiments | Total core-hours | Speedup vs. baseline |
|---|---|---|---|---|
| ROMS/MARBL (baseline) | 27,684 | 12 | 332,210 | -- |
| ROMS DOR + 2x12 mCDR response tracers | 16,333 | 1 | 16,333 | 20.3x |
| ROMS OAE + 2x12 mCDR response tracers | 18,699 | 1 | 18,699 | 17.8x |
| CESM/MARBL (baseline) | 422.8 | 690 | 291,701 | -- |
| CESM DOR + 1x690 mCDR response tracers | 10,995 | 1 | 10,995 | 26.5x |
| CESM OAE + 1x690 mCDR response tracers | 22,890 | 1 | 22,890 | 12.7x |

Note: the CESM baseline is now averaged over 2 samples (`control.001_backup`, no CDR forcing, and `glb-dor_North_Atlantic_basin_000...`, a real single-site DOR-forced full-BGC run) instead of 1. The two agree reasonably well (446.3 vs. 399.2 pe-hrs/simulated_year), supporting the assumption that per-site baseline cost is roughly forcing-independent, same as observed for ROMS. `smyle.cdr-atlas-v0.control.001` (without `_backup`) still has no completed timing data. As with ROMS, DOR- and OAE-forced tracer runs are kept as separate rows in both models since the cost gap between them is systematic (~1.14x for ROMS, ~2.1x for CESM), not noise.